File: services/shuttle-v2/scripts/eta-replay/general-eval/models/learned_distribution.py
"""Atom-preserving total-stand survival law used by the coherent challenger.

This is the analytic candidate's left/right quantile-knot convention with
log-survival arithmetic, so an overdue exponential tail does not underflow.
"""
import math
import numpy as np

from learned_features import LEVELS


def atom_remaining_quantiles(values, elapsed, levels=LEVELS):
    values = np.maximum(0, np.sort(np.asarray(values, dtype=float)))
    if not len(values) or not np.isfinite(values).all():
        raise ValueError('Finite nonempty quantiles required')
    xs, left, right = [], [], []
    if values[0] > 0:
        xs.append(0.)
        left.append(0.)
        right.append(0.)
    for i, value in enumerate(values):
        probability = (i + .5) / len(values)
        if xs and value == xs[-1]:
            right[-1] = probability
        else:
            xs.append(float(value))
            left.append(probability)
            right.append(probability)
    width = values[-1] - values[-2] if len(values) > 1 else 0
    hazard = min(1 / 5, max(1 / 1800, math.log(3) / width if width > 0 else 1 / 5))
    log_left = np.log1p(-np.asarray(left))
    log_right = np.log1p(-np.asarray(right))
    rest = max(0., float(elapsed))
    last = len(xs) - 1
    if rest >= xs[last]:
        # Beyond the final knot, survival conditioning is exactly memoryless.
        # Returning directly also avoids cancellation at very large rest ages.
        return -np.log1p(-np.asarray(levels)) / hazard
    index = 0
    while index < last and xs[index + 1] <= rest:
        index += 1
    if rest == xs[index]:
        log_survival = log_right[index]
    else:
        fraction = (rest - xs[index]) / (xs[index + 1] - xs[index])
        log_survival = log_right[index] + fraction * (log_left[index + 1] - log_right[index])
    result = []
    for level in levels:
        target = log_survival + math.log1p(-float(level))
        total = None
        for i in range(last + 1):
            if target >= log_left[i]:
                if i == 0:
                    total = xs[0]
                elif log_left[i] == log_right[i - 1]:
                    total = xs[i]
                else:
                    fraction = (target - log_right[i - 1]) / (log_left[i] - log_right[i - 1])
                    total = xs[i - 1] + fraction * (xs[i] - xs[i - 1])
                break
            if target >= log_right[i]:
                total = xs[i]
                break
        if total is None:
            total = xs[last] + (log_right[last] - target) / hazard
        result.append(max(0., total - rest))
    return np.asarray(result)

File: services/shuttle-v2/scripts/eta-replay/general-eval/models/learned-distribution.ts
/** Atom-preserving total-duration law, matching learned_distribution.py. */
export function atomRemainingQuantiles(values: readonly number[], elapsed: number,
  levels: readonly number[] = Array.from({ length: 19 }, (_, i) => (i + .5) / 19)): number[] {
  const q = values.map(x => Math.max(0, x)).sort((a, b) => a - b);
  if (!q.length || q.some(x => !Number.isFinite(x))) throw new Error("Finite nonempty quantiles required");
  const xs: number[] = [], left: number[] = [], right: number[] = [];
  if (q[0]! > 0) { xs.push(0); left.push(0); right.push(0); }
  for (let i = 0; i < q.length; i++) {
    const x = q[i]!, p = (i + .5) / q.length;
    if (xs.length && x === xs[xs.length - 1]) right[right.length - 1] = p;
    else { xs.push(x); left.push(p); right.push(p); }
  }
  const width = q.length > 1 ? q[q.length - 1]! - q[q.length - 2]! : 0;
  const hazard = Math.min(1 / 5, Math.max(1 / 1800, width > 0 ? Math.log(3) / width : 1 / 5));
  const logLeft = left.map(x => Math.log1p(-x)), logRight = right.map(x => Math.log1p(-x));
  const rest = Math.max(0, elapsed), last = xs.length - 1;
  if (rest >= xs[last]!) return levels.map(p => -Math.log1p(-p) / hazard);
  let index = 0;
  while (index < last && xs[index + 1]! <= rest) index++;
  const logSurvival = rest === xs[index] ? logRight[index]! : logRight[index]! +
    (rest - xs[index]!) / (xs[index + 1]! - xs[index]!) * (logLeft[index + 1]! - logRight[index]!);
  return levels.map(p => {
    const target = logSurvival + Math.log1p(-p);
    for (let i = 0; i <= last; i++) {
      if (target >= logLeft[i]!) {
        const total = i === 0 ? xs[0]! : logLeft[i] === logRight[i - 1] ? xs[i]! : xs[i - 1]! +
          (target - logRight[i - 1]!) / (logLeft[i]! - logRight[i - 1]!) * (xs[i]! - xs[i - 1]!);
        return Math.max(0, total - rest);
      }
      if (target >= logRight[i]!) return Math.max(0, xs[i]! - rest);
    }
    return Math.max(0, xs[last]! + (logRight[last]! - target) / hazard - rest);
  });
}
